src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/ArraySinglePrimitiveRecord.cs

```diff
@@ -145,6 +163,29 @@ internal static IReadOnlyList<T> DecodePrimitiveTypes(BinaryReader reader, int c
             }
         }
 
+        if (typeof(T) == typeof(bool))
+        {
+            // See DontCastBytesToBooleans test to see what could go wrong.
+            bool[] booleans = (bool[])(object)result;
+            for (int i = 0; i < booleans.Length; i++)
+            {
+                if (booleans[i]) // it can be any byte different than 0
+                {
+                    booleans[i] = true; // set it to 1 in explicit way
+                }
```

Comment:
```cs
                if (booleans[i]) // it can be any byte different than 0
                {
                    booleans[i] = true; // set it to 1 in explicit way
                }
```

It feels like an optimizing compiler or JIT could elide this.

If you want to perform fixup after the fact, my recommendation is to reinterpret_cast the whole thing back to a `Span<byte>` and check the elements _as bytes_. Anything that is not a 0 should be set forcibly set to 1.